pkg/export/setup/setup.go

```diff
@@ -82,138 +83,147 @@ func SetGlobal(exporter *export.Exporter) (err error) {
+		UintVar(&opts.Efficiency.ShardBufferSize)
 
 	a.Flag("export.token-url", "The request URL to generate token that's needed to ingest metrics to the project").
+		Default(opts.TokenURL).
 		StringVar(&opts.TokenURL)
 
 	a.Flag("export.token-body", "The request Body to generate token that's needed to ingest metrics to the project.").
+		Default(opts.TokenBody).
 		StringVar(&opts.TokenBody)
 
 	a.Flag("export.quota-project", "The projectID of an alternative project for quota attribution.").
+		Default(opts.QuotaProject).
 		StringVar(&opts.QuotaProject)
+}
 
-	haBackend := a.Flag("export.ha.backend", fmt.Sprintf("Which backend to use to coordinate HA pairs that both send metric data to the GCM API. Valid values are %q or %q", HABackendNone, HABackendKubernetes)).
-		Default(HABackendNone).Enum(HABackendNone, HABackendKubernetes)
-
-	kubeConfigPath := a.Flag("export.ha.kube.config", "Path to kube config file.").
-		Default("").String()
-	kubeNamespace := a.Flag("export.ha.kube.namespace", "Namespace for the HA locking resource. Must be identical across replicas. May be set through the KUBE_NAMESPACE environment variable.").
-		Default("").OverrideDefaultFromEnvar("KUBE_NAMESPACE").String()
-	kubeName := a.Flag("export.ha.kube.name", "Name for the HA locking resource. Must be identical across replicas. May be set through the KUBE_NAME environment variable.").
-		Default("").OverrideDefaultFromEnvar("KUBE_NAME").String()
-
-	// NOTE(bwplotka): This function will be likely performed within "getting ready" period, so before readiness is
-	// set to ready. Typical readiness can be as fast as 30s, so make sure this code timeouts faster than that.
-	return func(ctx context.Context, logger log.Logger, metrics prometheus.Registerer) (*export.Exporter, error) {
-		_ = level.Debug(logger).Log("msg", "started constructing the GCM export logic")
-
-		if metadata.OnGCE() {
-			// NOTE: OnGCE does not guarantee we will have all metadata entries or metadata
-			// server is accessible.
-
-			_ = level.Debug(logger).Log("msg", "detected we might run on GCE node; attempting metadata server access", "timeout", metadataFetchTimeout.String())
-			// When, potentially, on GCE we attempt to populate some, unspecified option entries
-			// like project ID, cluster, location, zone and user agent from GCP metadata server.
-			//
-			// This will be used to get *some* data, if not specified override by flags, to
-			// use if labels or external label settings does not have those set. Those will
-			// be used as crucial labels for export to work against GCM's Prometheus target.
-			//
-			// Set a hard time limit due to readiness and liveliness probes during this stage.
-			mctx, cancel := context.WithTimeout(context.Background(), metadataFetchTimeout)
-			tryPopulateUnspecifiedFromMetadata(mctx, logger, &opts)
-			cancel()
-			_ = level.Debug(logger).Log("msg", "best-effort on-GCE metadata gathering finished")
-		}
+type HAOptions struct {
```

Comment:
question: Why separate the haBackend config?

And either way, if want to do that, it would be much easier to review in a dedicated PR with a justification provided as there's a lot going on here already :)

Reply:
Addressing in https://github.com/GoogleCloudPlatform/prometheus-engine/pull/1001